### PR TITLE
MISC: Check that min versions are aligned in CI and import_optional_dependency

### PR DESCRIPTION
**Pull request by mroeschke**
Follow up to include `install.rst`

### CODE REVIEW

**Comment by attack68** [comment]
when you update pre-commit hooks, does one have to update dev environment with some git command, or is the addition of new files and new code enough for the hooks to operate as expected?

File: .pre-commit-config.yaml [hunk 1]
```diff
@@ -181,3 +181,7 @@ repos:
         entry: 'pg8000'
         files: ^ci/deps
         types: [yaml]
+    -   id: validate-min-versions-in-sync
```
Comment (lithomas1) [inline .pre-commit-config.yaml:184]: Can we combine this hook with the ``pg8000-not-installed-CI`` and ``pip-to-conda``?

**Review by lithomas1 — CHANGES_REQUESTED** [review]

**Comment by mroeschke** [comment]
> when you update pre-commit hooks, does one have to update dev environment with some git command, or is the addition of new files and new code enough for the hooks to operate as expected?

Nope. pre-commit will run (and install new dependencies if needed) with the changes of `.pre-commit-config.yaml` and addition of scripts. IIRC pre-commit runs in it's _own_ virtual environment.

File: .pre-commit-config.yaml (hunk 1 above)
Comment (mroeschke) [reply .pre-commit-config.yaml:184]: These 3 checks are independent though all dependency related, so I'd prefer to have these separate.

`pip-to-conda` checks local development dependencies

`pg8000-not-installed-CI` checks all CI deps

`validate-min-versions-in-sync` checks min versions across docs (future), CI, and code

Comment (lithomas1) [reply .pre-commit-config.yaml:191]: Ok, thats fine. I was just concerned about the growing amount of pre-commit hooks.

File: pandas/compat/_optional.py [hunk 2]
```diff
@@ -59,7 +66,8 @@ def get_version(module: types.ModuleType) -> str:
 
     if version is None:
         raise ImportError(f"Can't determine version for {module.__name__}")
-    return version
+    # psycopg2 appends " (dt dec pq3 ext lo64)" to it's version
+    return version.split()[0]
```
Comment (lithomas1) [inline pandas/compat/_optional.py:70]: Can you special case this for psycopg2 only?

File: scripts/validate_min_versions_in_sync.py [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+import sys
+
+DOC_PATH = pathlib.Path("doc/source/getting_started/install.rst").resolve()
+CI_PATH = next(
+    pathlib.Path("ci/deps").absolute().glob("actions-*-minimum_versions.yaml")
+)
+CODE_PATH = pathlib.Path("pandas/compat/_optional.py").resolve()
+
+
+def get_versions_from_code(content: str) -> dict[str, str]:
```
Comment (lithomas1) [inline scripts/validate_min_versions_in_sync.py:27]: Hm. Instead of using ast, can we just import the dict(pandas.compat._optional.VERSION)?

Comment (mroeschke) [reply scripts/validate_min_versions_in_sync.py:27]: When I try to add `from pandas.compat._optional import VERSIONS`, I get

```
% pre-commit run validate-min-versions-in-sync --all-files
Check minimum version of dependencies are aligned........................Failed
- hook id: validate-min-versions-in-sync
- exit code: 1

Traceback (most recent call last):
  File "scripts/validate_min_versions_in_sync.py", line 20, in <module>
    from pandas.compat._optional import VERSIONS
ModuleNotFoundError: No module named 'pandas'
Traceback (most recent call last):
  File "scripts/validate_min_versions_in_sync.py", line 20, in <module>
    from pandas.compat._optional import VERSIONS
ModuleNotFoundError: No module named 'pandas'
Traceback (most recent call last):
  File "scripts/validate_min_versions_in_sync.py", line 20, in <module>
    from pandas.compat._optional import VERSIONS
ModuleNotFoundError: No module named 'pandas'
Traceback (most recent call last):
  File "scripts/validate_min_versions_in_sync.py", line 20, in <module>
    from pandas.compat._optional import VERSIONS
ModuleNotFoundError: No module named 'pandas'
```

I think because these pre-commit hooks run in their own venv, they are not aware of the pandas directory as package

File: pandas/compat/_optional.py (hunk 2 above)
Comment (mroeschke) [reply pandas/compat/_optional.py]: Sure, changed.

File: .pre-commit-config.yaml (hunk 1 above)
Comment (mroeschke) [reply .pre-commit-config.yaml:191]: Fair concern. Might be good to eventually evaluate if some of these can just be CI runtime checks.

File: scripts/validate_min_versions_in_sync.py (hunk 3 above)
Comment (lithomas1) [reply scripts/validate_min_versions_in_sync.py:27]: Interesting. Maybe try to add the compat dir to ``sys.path``?
e.g.
``
>>> import sys
>>> sys.path.append("pandas/compat")
>>> import _optional
>>> _optional.VERSIONS
``

Comment (mroeschke) [reply scripts/validate_min_versions_in_sync.py:27]: So with this change

```
% git diff
diff --git a/scripts/validate_min_versions_in_sync.py b/scripts/validate_min_versions_in_sync.py
index 19728061e8..f6c03c099d 100755
--- a/scripts/validate_min_versions_in_sync.py
+++ b/scripts/validate_min_versions_in_sync.py
@@ -17,6 +17,11 @@ import ast
 import pathlib
 import sys

+sys.path.append("pandas/compat")
+import _optional
+
+version = _optional.VERSIONS
+
```

I get

```
% pre-commit run validate-min-versions-in-sync --all-files
Check minimum version of dependencies are aligned........................Failed
- hook id: validate-min-versions-in-sync
- exit code: 1

Traceback (most recent call last):
  File "scripts/validate_min_versions_in_sync.py", line 21, in <module>
    import _optional
  File "pandas/compat/_optional.py", line 8, in <module>
    from pandas.util.version import Version
ModuleNotFoundError: No module named 'pandas'
```

Comment (lithomas1) [reply scripts/validate_min_versions_in_sync.py:27]: Sorry about the error and for pushing on this issue. I've pulled your PR locally and I think I've found something that works. I'll add it as a suggestion.

File: scripts/validate_min_versions_in_sync.py [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+import pathlib
+import sys
+
+DOC_PATH = pathlib.Path("doc/source/getting_started/install.rst").resolve()
+CI_PATH = next(
+    pathlib.Path("ci/deps").absolute().glob("actions-*-minimum_versions.yaml")
+)
+CODE_PATH = pathlib.Path("pandas/compat/_optional.py").resolve()
+
+
```
Comment (lithomas1) [inline scripts/validate_min_versions_in_sync.py:26]:
```suggestion
# pandas package is not available
# in pre-commit environment
sys.path.append("pandas/compat")
sys.path.append("pandas/util")
import version
sys.modules["pandas.util.version"] = version

import _optional

versions = _optional.VERSIONS
```

File: scripts/validate_min_versions_in_sync.py (hunk 3 above)
Comment (mroeschke) [reply scripts/validate_min_versions_in_sync.py]: Great that worked! I agree this approach is better than parsing ASTs

**Review by lithomas1 — APPROVED** [review]
LGTM. Someone more familiar with the pre-commit hooks might want to review this though.
cc @MarcoGorelli.

File: .pre-commit-config.yaml [hunk 5]
```diff
@@ -188,3 +188,7 @@ repos:
...
```
Comment (MarcoGorelli) [inline .pre-commit-config.yaml:191]: looks like this hook only runs on certain files? if so, you can put `files: `, so it doesn't need to be run if only unrelated files are modified

**Review by MarcoGorelli — CHANGES_REQUESTED** [review]
generally looks good, just left a comment on `files`

File: .pre-commit-config.yaml (hunk 5 above)
Comment (mroeschke) [reply .pre-commit-config.yaml:191]: Good call. Added the files that should be checked.

File: scripts/validate_min_versions_in_sync.py [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+def main():
+    with open(CI_PATH) as f:
+        _, ci_optional = get_versions_from_ci(f.readlines())
+    code_optional = get_versions_from_code()
+    diff = set(ci_optional.items()).symmetric_difference(code_optional.items())
+    if diff:
+        sys.stdout.write(
+            f"The follow minimum version differences were found between  "
+            f"{CI_PATH} and {CODE_PATH}. Please ensure these are aligned: "
+            f"{diff}"
```
Comment (MarcoGorelli) [inline scripts/validate_min_versions_in_sync.py:74]:
```suggestion
            f"{diff}\n"
```

File: .pre-commit-config.yaml [hunk 7]
```diff
@@ -188,3 +188,8 @@ repos:
...
+        name: Check minimum version of dependencies are aligned
+        entry: python scripts/validate_min_versions_in_sync.py
+        language: python
+        files: ^(ci/deps/actions-*-minimum_versions\.yaml|pandas/compat/_optional\.py)$
```
Comment (MarcoGorelli) [inline .pre-commit-config.yaml:195]: This is regex (not glob), I presume you meant
```suggestion
        files: ^(ci/deps/actions-.*-minimum_versions\.yaml|pandas/compat/_optional\.py)$
```
?

File: scripts/validate_min_versions_in_sync.py [hunk 8]
```diff
@@ -0,0 +1,81 @@
...
+            package = package[2:]
+            if not seen_optional:
+                required_deps[package] = version
+            else:
+                optional_deps[package] = version
+    return required_deps, optional_deps
+
+
+def main():
+    with open(CI_PATH) as f:
```
Comment (MarcoGorelli) [inline scripts/validate_min_versions_in_sync.py:66]:
```suggestion
    with open(CI_PATH, encoding="utf-8") as f:
```
to prevent potential errors from people running Windows

**Review by MarcoGorelli — CHANGES_REQUESTED** [review]

File: pandas/compat/_optional.py [hunk 9]
```diff
@@ -59,6 +66,9 @@ def get_version(module: types.ModuleType) -> str:
 
     if version is None:
         raise ImportError(f"Can't determine version for {module.__name__}")
+    if module.__name__ == "psycopg2":
```
Comment (jreback) [inline pandas/compat/_optional.py:69]: can you integrate this to L64, e.g. maybe make a separate make of module -> lambda functions that can handle the special case modules?

File: scripts/validate_min_versions_in_sync.py [hunk 10]
```diff
@@ -0,0 +1,81 @@
...
+)
+CODE_PATH = pathlib.Path("pandas/compat/_optional.py").resolve()
+# pandas package is not available
+# in pre-commit environment
+sys.path.append("pandas/compat")
+sys.path.append("pandas/util")
+import version
+
+sys.modules["pandas.util.version"] = version
+import _optional
```
Comment (jreback) [inline scripts/validate_min_versions_in_sync.py:31]: can you import these as fully qualified and/or make this whole section a bit more idiomatic (i know tricky, but give a try)

**Review by jreback — CHANGES_REQUESTED** [review]

**Comment by jreback** [comment]
also pls rebase

File: pandas/compat/_optional.py (hunk 9 above)
Comment (mroeschke) [reply pandas/compat/_optional.py:69]: Hmm the L64 block handles if `__version__` is not the correct path to the version. psycopg2 uses `__version__` so it wouldn't hit this path

File: scripts/validate_min_versions_in_sync.py (hunk 10 above)
Comment (mroeschke) [reply scripts/validate_min_versions_in_sync.py:31]: Tried playing around with this but couldn't simplify it further than what @lithomas1 suggested.

**Comment by jreback** [comment]
@mroeschke can you rebase this to make sure ok (as merged @lithomas1 change)

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
https://dev.azure.com/pandas-dev/pandas/_build/results?buildId=72445&view=logs&j=e2350215-d973-503e-a41e-d287ff823797&t=751e993b-1954-5330-832a-a54bd1064f7d are these failing on master?

**Comment by mroeschke** [comment]
> https://dev.azure.com/pandas-dev/pandas/_build/results?buildId=72445&view=logs&j=e2350215-d973-503e-a41e-d287ff823797&t=751e993b-1954-5330-832a-a54bd1064f7d are these failing on master?

Yes it appears so. 

https://dev.azure.com/pandas-dev/pandas/_build/results?buildId=72425&view=logs&j=e2350215-d973-503e-a41e-d287ff823797&t=751e993b-1954-5330-832a-a54bd1064f7d&l=44

**Comment by simonjayhawkins** [comment]
it's not urgent (i.e. block 1.4) but would be good to keep the ci files in sync until they need to diverge for say Python 3.11 so that ci backports have less chance of needing manual intervention.

could either sync ci/deps/actions-38-minimum_versions.yaml or backport this PR.

**Comment by mroeschke** [comment]
Might be better to backport this PR since it has some additional changes to `pandas/compat/_optional.py`, but no strong opinion on my side.

**Comment by simonjayhawkins** [comment]
@meeseeksdev backport 1.4.x
